presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanSourceProvider.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.sql.planner.PlanFragment;
+import com.facebook.presto.sql.planner.plan.PlanFragmentId;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.RemoteSourceNode;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+public class FragmentedPlanSourceProvider
+        implements PlanNodeSourceProvider
+{
+    private Map<PlanFragmentId, PlanNode> fragmentPlanNodeMap;
+
+    private FragmentedPlanSourceProvider(Map<PlanFragmentId, PlanNode> fragmentPlanNodeMap)
+    {
+        this.fragmentPlanNodeMap = requireNonNull(fragmentPlanNodeMap, "fragmentPlanNodeMap is null");
+    }
+
+    @Override
+    public List<PlanNode> getSources(PlanNode node)
+    {
+        if (node instanceof RemoteSourceNode) {
+            return ((RemoteSourceNode) node).getSourceFragmentIds().stream()
+                    .map(id -> {
+                        verify(fragmentPlanNodeMap.containsKey(id), "fragment id not in map");
+                        return fragmentPlanNodeMap.get(id);
+                    })
+                    .collect(toImmutableList());
+        }
+
+        return node.getSources();
+    }
+
+    public static FragmentedPlanSourceProvider createFragmentedPlanSourceProvider(List<PlanFragment> planFragments)
```

Comment:
`s/createFragmentedPlanSourceProvider/create`

Also I would put it at the top, above the constructor